crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
+}
+
+impl Ranged for DunderAllValue {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+#[derive(Debug)]
+enum SortedDunderAll {
+    AlreadySorted,
+    Sorted(Option<String>),
+}
+
+fn collect_dunder_all_lines(range: TextRange, locator: &Locator) -> Option<Vec<DunderAllLine>> {
+    // Collect data on each line of `__all__`.
+    // Return `None` if `__all__` appears to be invalid,
+    // or if it's an edge case we don't care about.
+    let mut parentheses_open = false;
+    let mut lines = vec![];
+    let mut items_in_line = vec![];
+    let mut comment_in_line = None;
+    for pair in lexer::lex(locator.slice(range).trim(), Mode::Expression) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
+            }
+            Tok::Rpar | Tok::Rsqb | Tok::Newline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                    items_in_line.clear();
+                    comment_in_line = None;
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
+            Tok::String { value, .. } => items_in_line.push((value, subrange)),
+            Tok::Comma => continue,
+            _ => return None,
+        }
+    }
+    Some(lines)
+}
+
+#[derive(Debug)]
+struct DunderAllLine {
+    items: Vec<(String, TextRange)>,
+    comment: Option<TextRange>,
+}
+
+impl DunderAllLine {
+    fn new(items: &[(String, TextRange)], comment: Option<TextRange>, offset: TextSize) -> Self {
+        assert!(comment.is_some() || !items.is_empty());
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
```

Comment:
Nit: I have a preference towards avoiding single-letter variables in favor of more explicit names. E.g., `item` and `range` here, maybe?

Reply:
I think we should be able to avoid this `.to_owned()`... E.g., could we pass in `&mut Vec`? Or maybe we pass in `items.drain(..)` as the argument?

Reply:
@MichaReiser and I vanquished all uses of `.clone()` in a pair-programming session; this region of code no longer exists :)